File: Execution/affordance_corrections/nodes/affordance_corrections/affordance_helpers/urdf_helpers.py
#!/usr/bin/env python3
# File name: urdf_helpers.py
# Description: URDF to class objects
# Author: Mike Hagenow
# Date: 7/07/2021

import numpy as np
import trimesh
import xml.etree.ElementTree as ET
from scipy.spatial.transform import Rotation as ScipyR
import rospkg

class Joint:
    ''' used in model'''
    def __init__(self,axis,min_angle,max_angle,R,t):
        # store and sample meshes
        self.axis = axis
        self.min_angle = min_angle
        self.max_angle = max_angle
        self.R = R
        self.t = t

class Link:
    ''' used in model'''
    def __init__(self,meshfile,R,t, num_pts = 400):
        self.meshfile = meshfile
        self.R = R
        self.t = t

        self.mesh = trimesh.load_mesh(meshfile)
        self.pts = trimesh.sample.sample_surface_even(self.mesh,num_pts,radius=None)[0]

def getLinksAndJoints(urdffile):
    ''' From the URDF, load link and joint objects for fitting, plotting, etc.
    Note: an object without articulations would be expected to have one link'''
    
    xmltree = ET.parse(urdffile)
    root = xmltree.getroot()
    
    joints = []
    links = []

    # Determine number of links a priori to determine number of points per link
    # (same total number of points)
    num_links = 0
    for child in root:
        if child.tag == "link":
            num_links += 1
    if num_links == 0:
        num_links = 1  
    
    # The following is mainly traversing the XML structure
    # to pull out key information needed to define the full mesh
    for child in root:
        # TODO: only supports revolute currently
        if child.tag=="joint" and child.attrib['type']=='revolute':
            for attr in child:
                if attr.tag == "limit":
                    min_angle = float(attr.attrib['lower'])
                    max_angle = float(attr.attrib['upper'])
                if attr.tag == "axis":
                    axis = np.fromstring(attr.attrib['xyz'],dtype=float, sep=' ')
                    axis = axis/np.linalg.norm(axis) # normalize if necessary
                if attr.tag == "origin":
                    rpy = np.fromstring(attr.attrib['rpy'],dtype=float, sep=' ')
                    R = ScipyR.from_euler('zyx',rpy,degrees=False).as_matrix()
                    t = np.fromstring(attr.attrib['xyz'],dtype=float, sep=' ')
            joint_temp = Joint(axis,min_angle,max_angle,R,t)
            joints.append(joint_temp)
        if child.tag=="link":
            visual = child[0]
            for attr in visual:
                if attr.tag== "origin":
                    rpy = np.fromstring(attr.attrib['rpy'],dtype=float, sep=' ')
                    R = ScipyR.from_euler('zyx',rpy,degrees=False).as_matrix()
                    t = np.fromstring(attr.attrib['xyz'],dtype=float, sep=' ')
                if attr.tag == "geometry":
                    for geo_attr in attr:
                        if geo_attr.tag == "mesh":
                            mesh_file = geo_attr.attrib['filename']
                            # resolve full filename
                            pkg_temp = mesh_file.split("//")[1].split("/")[0]

                            rospack = rospkg.RosPack()
                            path_within_package = ""
                            for ii in range(1,len(mesh_file.split("//")[1].split("/"))):
                                path_within_package+=("/"+mesh_file.split("//")[1].split("/")[ii])
                            mesh_file_full = rospack.get_path(pkg_temp)+path_within_package
            link_temp = Link(mesh_file_full,R,t, num_pts =int(400.0/num_links))
            links.append(link_temp)

    return links, joints

def getLinksAndJointsFromSTL(stlfile):
    ''' An STL file can be interpreted as a single link'''
    joints = []
    links = []

    link_temp = Link(stlfile,np.eye(3),np.zeros((3,)))
    links.append(link_temp)

    return links, joints

def getPointsWithAngles(model,angles,R_base,t_base,scale=1):
    ''' gets the full point sampled model for a model with given joint angles'''
    pts_arr = np.zeros((0,3))

    for ii in range(0,len(model.links)):
        pts_temp = np.copy(model.links[ii].pts)
        # scale part
        centroid_before = np.mean(pts_temp,axis=0)
        pts_temp*=scale
        centroid_after = np.mean(pts_temp,axis = 0)
        pts_temp-=(centroid_after-centroid_before)

        for jj in range(ii,-1,-1): # count from ii down to 0

            # link offset (needs to be subtracted)
            pts_temp = (model.links[jj].R @ pts_temp.T).T - scale*model.links[jj].t
            
            # joint transform
            if (jj<len(model.joints) and ii!=0): # need to ignore final link (has no joint)
                # Transformation of joint itself
                
                pts_temp = (model.joints[jj].R @ pts_temp.T).T + scale*model.joints[jj].t

                # Rotation from the angle (with respect to the model's origin)
                R_angle = ScipyR.from_rotvec(angles[jj]*model.joints[jj].axis).as_matrix()
                # centroid_before = np.mean(pts_temp,axis=0)
                origin_before = -scale*model.links[jj].t
                origin_after = (R_angle @ origin_before.T).T
                pts_temp = (R_angle @ pts_temp.T).T
                pts_temp -= (origin_after-origin_before)
                
        # Rotation of entire structure (i.e., initial)
        pts_temp = (R_base @ pts_temp.T).T + t_base
        pts_arr = np.concatenate((pts_arr,pts_temp),axis=0)
    
    return pts_arr

File: Execution/affordance_corrections/nodes/affordance_corrections/testcode/urdf_helpers.cpp


#include </opt/ros/noetic/include/ros/package.h>
 
#include <string>
#include <regex>
#include <string>
#include <stdbool.h>
#include <list>
#include <tuple>
#include "NumCpp.hpp"


using namespace std; 
//using namespace trimesh;
class Joint{
       // ''' used in model'''
        public:
            nc::NdArray<float> axis;
            float min_angle;
            float max_angle;
            nc::NdArray<float> R; //don't know what type to put for R
            nc::NdArray<float> t; 
            Joint(nc::NdArray<float> axis,float min_angle,float max_angle,nc::NdArray<float> R,nc::NdArray<float> t);
};
Joint::Joint(nc::NdArray<float> axis,float min_angle,float max_angle,nc::NdArray<float> R,nc::NdArray<float> t){
        this->axis = axis;
        this->min_angle = max_angle;
        this->max_angle = max_angle;
        this->R = R;
        this->t = t;
}

class Link{
   // ''' used in model'''
    public:
        nc::NdArray<float> pts;
        nc::NdArray<float> R;
        nc::NdArray<float> t;
        Link(nc::NdArray<float> mesh_points, nc::NdArray<float> R, nc::NdArray<float> t, int num_pts);
};
Link::Link(nc::NdArray<float> mesh_points, nc::NdArray<float> R, nc::NdArray<float> t, int num_pts){

    this->R = R;
    this->t = t;
    this->pts = mesh_points;
}

//will be using the original python methods for this function

/*
static  std::tuple<std::vector<Link>, std::vector<Joint>> getLinksAndJoints(nc::NdArray<float> mesh_points, std::string urdffile){
    //''' From the URDF, load link and joint objects for fitting, plotting, etc.
    //Note: an object without articulations would be expected to have one link'''
     nc::NdArray<float> axis;
     nc::NdArray<float> R;
     nc::NdArray<float> t;
     float min_angle;
     float max_angle;
     std::string mesh_file_full;
    rapidxml::xml_document<> xmltree;
    //have to xmltree in buffer!!!!
    char *urdffile_param;
    strcpy(urdffile_param, urdffile.c_str());
    
    xmltree.parse<0>(urdffile_param); //should represent the root after this line
    rapidxml::xml_node<> * root = xmltree.first_node();
    std::vector<Link> links;
    std::vector<Joint> joints;
   

    //The following is mainly traversing the XML structure
    // to pull out key information needed to define the full mesh
    //check for 0 returns!!!!!
    for(rapidxml::xml_node<> *child = root->first_node(); child; child = child->next_sibling()){
         //TODO: only supports revolute currently
         std::string childTag (child->name());
         std::string childAttribute(child->first_attribute("type")->value());
        if((childTag.compare("joint") == 0) && (childAttribute.compare("revolute") == 0)){
            for(rapidxml::xml_node<> *attr = child->first_node(); attr; attr = attr->next_sibling()){
                std::string attrTag (attr->name());
                std::string::size_type size;
                if(attrTag.compare("limit") == 0){
                    min_angle = std::stof((attr->first_attribute("lower")->value()), &size );
                    max_angle = std::stof((attr->first_attribute("upper")->value()), &size );
                }
                if(attrTag.compare("axis") == 0){
                    std::string string = attr->first_attribute("xyz")->value();
                    int index;
                    std::vector<float> tempAxis; 
                    float coordin_float;
                    while((index = string.find(' ')) != std::string::npos){
                        std::string coordin = string.substr(0,index);
                        coordin_float = std::stof(coordin);
                        //add to vector then create ndarray? 
                        tempAxis.push_back(coordin_float);
                        string.erase(0, index + 1);
                    }
                    //need to get z coordinate 
                    coordin_float = std::stof(string);
                    tempAxis.push_back(coordin_float);
                    axis = nc::NdArray<float>(tempAxis);
                    auto axis_double = nc::norm(axis);  //normalize if necessary
                    axis = axis_double.astype<float>();
                }
                if(attrTag.compare("origin") == 0){
                    nc::NdArray<float> rpy; 
                    std::string string = attr->first_attribute("rpy")->value();
                    int index;
                    std::vector<float> tempRpy;
                    float coordin_float;
                    while((index = string.find(' ')) != std::string::npos){
                        std::string coordin = string.substr(0,index);
                        coordin_float = std::stof(coordin);
                        //add to vector then create ndarray? 
                        tempRpy.push_back(coordin_float);
                        string.erase(0, index + 1);
                    }
                    //need to get last coorindate 
                    coordin_float = std::stof(string);
                    tempRpy.push_back(coordin_float);
                    rpy = nc::NdArray<float>(tempRpy);
                    
                     nc::NdArray<float> R;
                    
                    R =  nc::eye<float>(3);
                   ////////////////////////////////////////////////////////////
                    std::string stringT = attr->first_attribute("xyz")->value();
                    int indexT;
                    std::vector<float> tempT;
                    float coordin_floatT;
                    while((indexT = stringT.find(' ')) != std::string::npos){
                        std::string coordin = string.substr(0,indexT);
                        coordin_floatT = std::stof(coordin);
                        //add to vector then create ndarray? 
                        tempT.push_back(coordin_floatT);
                        stringT.erase(0, index + 1);
                    }
                    //need to get z coordinate 
                    coordin_floatT = std::stof(stringT);
                    tempT.push_back(coordin_floatT);
                    auto t = nc::NdArray<float>(tempT);
                }
            }
            Joint joint_temp(axis,min_angle,max_angle,R,t);
            
            joints.push_back(joint_temp);
        }
        if(childTag.compare("link") == 0){
            rapidxml::xml_node<>* visual = (child->first_node("visual"));
            for(rapidxml::xml_node<> *attr = child->first_node(); attr; attr = attr->next_sibling()){
                std::string attrTag (attr->name());
                if(attrTag.compare("origin")){
                    nc::NdArray<float> rpy;
                    std::string string = attr->first_attribute("rpy")->value();
                    int index;
                    std::vector<float> tempRpy;
                    float coordin_float;
                    while((index = string.find(' ')) != std::string::npos){
                        std::string coordin = string.substr(0,index);
                        coordin_float = std::stof(coordin);
                        //add to vector then create ndarray? 
                        tempRpy.push_back(coordin_float);
                        string.erase(0, index + 1);
                    }
                    //need to get last coorindate 
                    coordin_float = std::stof(string);
                    tempRpy.push_back(coordin_float);
                    rpy = nc::NdArray<float>(tempRpy);

                    nc::NdArray<float> R;
                    R =  nc::eye<float>(3);
                   ////////////////////////////////////////////////////////////
                    std::string stringT = attr->first_attribute("xyz")->value();
                    int indexT;
                    std::vector<float> tempT;
                    float coordin_floatT;
                    while((indexT = stringT.find(' ')) != std::string::npos){
                        std::string coordin = string.substr(0,indexT);
                        coordin_floatT = std::stof(coordin);
                        //add to vector then create ndarray? 
                        tempT.push_back(coordin_floatT);
                        stringT.erase(0, index + 1);
                    }
                    //need to get z coordinate 
                    coordin_floatT = std::stof(stringT);
                    tempT.push_back(coordin_floatT);
                    auto t = nc::NdArray<float>(tempT);

                }
                if(attrTag.compare("geometry")){
                    for(rapidxml::xml_node<> *geo = child->first_node(); geo; geo = geo->next_sibling()){
                        std::string geo_attr_tag (geo->name());
                        if(geo_attr_tag.compare("mesh")){
                            std::string mesh_file (geo->first_attribute("filename")->value());
                            //resolve full filename
                            int double_slash_index = mesh_file.find_first_of("//");
                            std::string string_path = mesh_file.substr(double_slash_index);
                            int first_slash_index = string_path.find_first_of("/");
                            std::string path_within_package = "";
                            //Python code has for loop, but I don't think it is necessary
                            path_within_package = string_path;
                            std::string pkg_temp  = string_path.substr(0, first_slash_index);
                            //mesh_file_full = ros::package::getPath(pkg_temp)+ path_within_package;
                        }
                    }
                }
            }
            Link link_temp(mesh_points, R, t,400);
            links.push_back(link_temp);
        }
    }
    std::tuple<std::vector<Link>, std::vector<Joint>> links_joints (links, joints);
    return links_joints;
}

static std::tuple<std::vector<Link>, std::vector<Joint>>  getLinksAndJointsFromSTL(nc::NdArray<float> mesh_points, std::string stlfile){
    //''' An STL file can be interpreted as a single link'''
    std::vector<Link> links;
    std::vector<Joint> joints;
    
    nc::NdArray<float> R = nc::eye<float>(3); //creates diagonal matrix
    nc::NdArray<float> t = nc::zeros<float>(3);
    Link link_temp(mesh_points,R, t,400);
    links.push_back(link_temp);
    std::tuple<std::vector<Link>, std::vector<Joint>> links_joints (links, joints);
    return links_joints;
}*/
//2D array

